### ISSUE TITLE
decoupling 

### ISSUE THREAD

**Issue by EMalpha01** [EMalpha01/Sentence_VAE#1]
SentenceVAE/
│
├── encoder.py
│   ```python
│   import torch
│   from torch import nn
│
│   class SentenceEncoder(nn.Module):
│       '''Sentence Encoder with byte-level BPE tokenization, learned positional encoding, and validation checks'''
│       def __init__(self, model, tokenizer, hidden_size):
│           super(SentenceEncoder, self).__init__()
│           self.model = model
│           self.tokenizer = tokenizer
│           self.hidden_size = hidden_size
│           self.layer_norm = nn.LayerNorm(self.hidden_size)
│           if hasattr(self.model.decoder.embed_positions, 'offset'):
│               self.model.decoder.embed_positions.offset = 0
│
│       def forward(self, sentence):
│           # Encoding logic here
│           pass
│   ```
│
├── decoder.py
│   ```python
│   import torch
│   from torch import nn
│
│   class SentenceDecoder(nn.Module):
│       '''Sentence Decoder for generating sentences from sentence-level tokens, with validation checks'''
│       def __init__(self, model, tokenizer, hidden_size):
│           super(SentenceDecoder, self).__init__()
│           self.model = model
│           self.tokenizer = tokenizer
│           self.hidden_size = hidden_size
│
│       def forward(self, sentence_level_token):
│           # Decoding logic here
│           pass
│   ```
│
├── model.py
│   ```python
│   import torch
│   from torch import nn
│
│   class EncoderDecoderModel:
│       '''Encoder-Decoder Model with training logic, including focal loss and AdamW optimizer'''
│       def __init__(self, encoder, decoder, tokenizer, learning_rate=1e-4):
│           self.encoder = encoder
│           self.decoder = decoder
│           self.tokenizer = tokenizer
│           self.optimizer = torch.optim.AdamW(
│               list(self.encoder.parameters()) + list(self.decoder.parameters()), lr=learning_rate
│           )
│           self.device = "cuda" if torch.cuda.is_available() else "cpu"
│           if self.device == "cuda":
│               self.encoder = nn.DataParallel(self.encoder).to(self.device)
│               self.decoder = nn.DataParallel(self.decoder).to(self.device)
│           else:
│               self.encoder.to(self.device)
│               self.decoder.to(self.device)
│
│       def train(self, sentences):
│           # Training logic here
│           pass
│   ```
│
├── inference.py
│   ```python
│   import torch
│
│   class InferenceEngine:
│       '''Inference Engine for reconstructing text sentence-by-sentence'''
│       def __init__(self, encoder, decoder, tokenizer):
│           self.encoder = encoder
│           self.decoder = decoder
│           self.tokenizer = tokenizer
│
│       def inference(self, text):
│           # Inference logic here
│           pass
│   ```
│
├── utils.py
│   ```python
│   def split_into_sentences(text):
│       punctuation_marks = [".", "?", "!", ","]
│       sentences = []
│       current_sentence = ""
│       for char in text:
│           current_sentence += char
│           if char in punctuation_marks:
│               sentences.append(current_sentence.strip())
│               current_sentence = ""
│       if current_sentence:
│           sentences.append(current_sentence.strip())
│       return sentences
│   ```
│
├── config.py
│   ```python
│   MODEL_NAME = "facebook/opt-1.3b"
│   HIDDEN_SIZE = 2048
│   LEARNING_RATE = 1e-4
│   ```
│
├── main.py
│   ```python
│   import torch
│   from transformers import AutoTokenizer, AutoModel
│   from encoder import SentenceEncoder
│   from decoder import SentenceDecoder
│   from model import EncoderDecoderModel
│   from inference import InferenceEngine
│   from config import MODEL_NAME, HIDDEN_SIZE, LEARNING_RATE
│
│   def main():
│       tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
│       model = AutoModel.from_pretrained(MODEL_NAME)
│       encoder = SentenceEncoder(model, tokenizer, HIDDEN_SIZE)
│       decoder = SentenceDecoder(model, tokenizer, HIDDEN_SIZE)
│       encoder_decoder_model = EncoderDecoderModel(encoder, decoder, tokenizer, learning_rate=LEARNING_RATE)
│       sentences = ["Sample sentence for training."]
│       encoder_decoder_model.train(sentences)
│       inference_engine = InferenceEngine(encoder, decoder, tokenizer)
│       text = "Test for inference."
│       print("Reconstructed Text:", inference_engine.inference(text))
│
│   if __name__ == "__main__":
│       main()
│   ```
│
└── requirements.txt

**Comment by EMalpha01** [comment]
This project structure outlines a `SentenceVAE ` system with a clear separation of concerns for encoding, decoding, training, inference, configuration, and utility functions. Here’s a breakdown:

Directory Structure:
The `SentenceVAE ` folder organizes modules that work together to create an encoder-decoder model for sentence processing, likely geared towards reconstructing or generating sentences with a VAE (Variational Autoencoder) approach. Each module focuses on a specific part of the model’s lifecycle (encoding, decoding, training, and inference).

Files:

encoder.py:
Defines a `SentenceEncoder` class using PyTorch’s `nn.Module`, integrating a model with byte-level BPE tokenization, learned positional encoding, and optional validation checks. The encoder initializes with the main model and tokenizer and includes layer normalization to stabilize the hidden states. The forward method placeholder suggests encoding logic is to be implemented here.

decoder.py:
Implements `SentenceDecoder`, handling sentence generation from encoded tokens. Like the encoder, it’s initialized with the model, tokenizer, and hidden size, and its forward method will carry out the decoding logic.

model.py:
Contains `EncoderDecoderModel`, which brings together the encoder and decoder for training. It incorporates training logic, with a focal loss and the AdamW optimizer. The model adjusts for GPU usage if available, with `nn.DataParallel` applied for parallelism on CUDA-compatible devices.

inference.py:
Houses `InferenceEngine`, responsible for sentence-by-sentence reconstruction of text during inference. This class calls the encoder and decoder to generate output based on trained model weights.

utils.py:
Provides utility functions, such as `split_into_sentences`, which splits text based on punctuation marks (periods, question marks, exclamation points, and commas). This function is useful for parsing input text into sentences for encoding.

config.py:
Contains configuration parameters, such as `MODEL_NAME` (defaulting to `facebook/opt-1.3b`), `HIDDEN_SIZE` (2048), and `LEARNING_RATE`. This file helps centralize model and training configurations.

main.py:
The main script for running the model. It loads the tokenizer and model from the transformers library, initializes the encoder and decoder, and constructs the encoder-decoder model. After training on sample sentences, it uses the `InferenceEngine` for testing sentence reconstruction. This script allows direct execution of the entire model pipeline.

requirements.txt:
Expected to list dependencies, likely including `torch`, `transformers`, and any other libraries required for this project.

Summary: This structured project seems designed for experimentation with sentence-level VAE processing, offering components for tokenization, encoding, decoding, training, and inference, with a focus on sentence reconstruction or generation tasks using PyTorch and Transformers. The structure is modular, making it extensible and easy to modify individual components as needed.
